### PR TITLE
wire in aggregator

### PR DESCRIPTION
**Pull request by deads2k**
I think there are more picks that are going to be required and I'm sorting out some bad behavior around hangs, but this is the direction I'm going in.

@openshift/api-review for the the aggregator config commit.

Right now, I've only tested this manually (no test image is yet pushed and it requires custom master-config.yaml and active nodes).

1. Generate front-proxy-ca cert/key and aggregator-front-proxy-client cert/key.  I did this with
```bash
oadm ca create-signer-cert --cert=openshift.local.config/master/front-proxy-ca.crt --key=openshift.local.config/master/front-proxy-ca.key
oadm create-api-client-config --certificate-authority=openshift.local.config/master/front-proxy-ca.crt --signer-cert=openshift.local.config/master/front-proxy-ca.crt --signer-key=openshift.local.config/master/front-proxy-ca.key --user aggregator-front-proxy --client-dir=openshift.local.config/master
```
and then deleting everything but the files I wanted.
2. Generated a fresh master-config.yaml.
3. Replace stanzas in master-config.yaml with
```yaml
aggregatorConfig:
  proxyClientInfo:
    certFile: aggregator-front-proxy.crt
    keyFile: aggregator-front-proxy.key
authConfig:
  requestHeader:
    clientCA: front-proxy-ca.crt
    clientCommonNames: 
    - aggregator-front-proxy
    usernameHeaders:
    - X-Remote-User
    groupHeaders:
    - X-Remote-Group
    extraHeaderPrefixes:
    - X-Remote-Extra-
```
4. `sudo $(which openshift) start --master-config openshift.local.config/master/master-config.yaml --node-config openshift.local.config/node-deads-dev-01/node-config.yaml`
5. Create the image to use by cloning kube and building: `nice make WHAT=vendor/k8s.io/sample-apiserver/ && vendor/k8s.io/sample-apiserver/hack/build-image.sh`
6. `oc new-project wardle` - new project to start with
7. `oadm policy add-scc-to-user privileged -z apiserver` - etcd doesn't like running as non-root and I didn't feel like trying to deal with it
8. `oc create -f test/extended/testdata/aggregator/` - create the resources
9. `oc get flunders` - proof!

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
`Error from server (InternalError): an error on the server ("Error: 'net/http: invalid header field name \"X-Remote-Extra-authorization.openshift.io/scopes\"'\nTrying to reach: 'https://172.30.134.150/apis/wardle.k8s.io/v1alpha1'") has prevented the request from succeeding
`

Well, that's a problem.  @liggitt got suggestions?

**Comment by deads2k** [comment]
Well, it can proxy certificate based users at the moment, so its a strict improvement.  Let's get some reviews on this much.

**Comment by deads2k** [comment]
@bparees you were asking.

**Comment by liggitt** [comment]
>> Error from server (InternalError): an error on the server ("Error: 'net/http: invalid header field name \"X-Remote-Extra-authorization.openshift.io/scopes\"'\nTrying to reach: 'https://172.30.134.150/apis/wardle.k8s.io/v1alpha1'") has prevented the request from succeeding
> 
> Well, that's a problem. @liggitt got suggestions?

...

**Comment by deads2k** [comment]
> Well, that's a problem. @liggitt got suggestions?

Yeah.  Technically its never worked, so changing the name can work.  `scopes.authorization.openshift.io` I guess.

**Comment by deads2k** [comment]
I won't claim its turnkey, but I've added directions @bparees or @pmorie want to try it.

**Comment by bparees** [comment]
jessica actually gave me a path to console SC usage that doesn't require the aggregator after all, so i'm going to try that first.  I'm still bleeding from my experience as an api server guinea pig.

**Comment by deads2k** [comment]
@sdodson I'm guessing that @pmorie and @derekwaynecarr have already engaged you regarding how to wire up the service catalog for its beta installation, but the instructions I've included in the description show the particular fields used for wiring the API server to enable the aggregator to route traffic to the service catalog.

**Comment by sdodson** [comment]
@ewolinetz and @jpeeler are collaborating (or about to) on that work.

File: pkg/cmd/server/api/v1/types.go [hunk 1]
```diff
@@ -300,6 +303,12 @@ type RequestHeaderAuthenticationOptions struct {
 	ExtraHeaderPrefixes []string `json:"extraHeaderPrefixes"`
 }
 
+// AggregatorConfig holds information required to make the aggregator function.
+type AggregatorConfig struct {
+	// ProxyClientInfo specifies the client cert/key to use when proxying to aggregated API servers
+	ProxyClientInfo CertInfo `json:"proxyClientInfo"`
```
Comment (liggitt) [inline pkg/cmd/server/api/v1/types.go]: add to `ResolveMasterConfigPaths`

File: pkg/cmd/server/api/types.go [hunk 2]
```diff
@@ -436,6 +439,12 @@ type RequestHeaderAuthenticationOptions struct {
 	ExtraHeaderPrefixes []string
 }
 
+// AggregatorConfig holds information required to make the aggregator function.
+type AggregatorConfig struct {
+	// ProxyClientInfo specifies the client cert/key to use when proxying to aggregated API servers
+	ProxyClientInfo CertInfo
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: GetMasterFileReferences

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: nm, missed it at the top

**Comment by liggitt** [comment]
`add aggregator config` commit LGTM

File: pkg/cmd/server/origin/aggregator.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+		if len(tokens) != 4 {
+			continue
+		}
+
+		apiService := makeAPIService(schema.GroupVersion{Group: tokens[2], Version: tokens[3]})
+
+		// TODO this is probably an indication that we need explicit and precise control over the discovery chain
+		// but for now its a special case
+		// apps has to come last for compatibility with 1.5 kubectl clients
+		if apiService.Spec.Group == "apps" {
```
Comment (mfojtik) [inline pkg/cmd/server/origin/aggregator.go]: just to confirm, this won't conflict with `apps.openshift.io`?

File: pkg/cmd/server/origin/master.go [hunk 4]
```diff
@@ -193,7 +193,28 @@ func (c *MasterConfig) Run(kc *kubernetes.MasterConfig, assetConfig *AssetConfig
...
+		go apiserver.Run(utilwait.NeverStop)
+
+		// Attempt to verify the server came up for 20 seconds (100 tries * 100ms, 100ms timeout per try)
+		cmdutil.WaitForSuccessfulDial(c.TLS, c.Options.ServingInfo.BindNetwork, c.Options.ServingInfo.BindAddress, 100*time.Millisecond, 100*time.Millisecond, 100)
+		return
+	}
+
+	aggregatorConfig, err := c.createAggregatorConfig(*kc.Master.GenericConfig)
+	if err != nil {
+		glog.Fatalf("Failed to launch master: %v", err)
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master.go]: Failed to create aggregator config

File: pkg/authorization/api/types.go [hunk 5]
```diff
@@ -22,7 +22,7 @@ const (
 	VerbAll        = "*"
 	NonResourceAll = "*"
 
-	ScopesKey           = "authorization.openshift.io/scopes"
+	ScopesKey           = "scopes.authorization.openshift.io"
```
Comment (mfojtik) [inline pkg/authorization/api/types.go]: do we need to update/migrate something after this change?

File: pkg/cmd/server/origin/aggregator.go (hunk 3 above)
Comment (mfojtik) [inline pkg/cmd/server/origin/aggregator.go]: nvmd.

**Comment by mfojtik** [comment]
LGTM, but will prefer @enj to have second look before merging.

File: pkg/authorization/api/types.go (hunk 5 above)
Comment (deads2k) [inline pkg/authorization/api/types.go]: > do we need to update/migrate something after this change?

No.  I reasoned through this with jordan. It's merging in a separate pull.

**Comment by deads2k** [comment]
@smarterclayton I seem to be failing on 
```
===== Verifying Generated Bindata =====
Generating bindata...
FAILURE: Generation of fresh bindata failed:
error: extended bindata is 656537 bytes, reduce the size of the import
```

Any idea what's up?

**Comment by deads2k** [comment]
comments addressed [merge]

File: pkg/cmd/server/origin/master.go [hunk 6]
```diff
@@ -193,7 +193,28 @@ func (c *MasterConfig) Run(kc *kubernetes.MasterConfig, assetConfig *AssetConfig
...
+		// Attempt to verify the server came up for 20 seconds (100 tries * 100ms, 100ms timeout per try)
+		cmdutil.WaitForSuccessfulDial(c.TLS, c.Options.ServingInfo.BindNetwork, c.Options.ServingInfo.BindAddress, 100*time.Millisecond, 100*time.Millisecond, 100)
+		return
+	}
+
+	aggregatorConfig, err := c.createAggregatorConfig(*kc.Master.GenericConfig)
+	if err != nil {
+		glog.Fatalf("Failed to create aggregator config: %v", err)
+	}
+	aggregatorServer, err := createAggregatorServer(aggregatorConfig, apiserver.GenericAPIServer, kc.Informers.InternalKubernetesInformers(), stopCh)
```
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: > Do you plan to use kc.Informers.InternalKubernetesInformers() later?

it gets started elsewhere to drive this.

File: pkg/cmd/server/origin/master.go [hunk 7]
```diff
@@ -193,7 +193,28 @@ func (c *MasterConfig) Run(kc *kubernetes.MasterConfig, assetConfig *AssetConfig
...
-	go kmaster.GenericAPIServer.PrepareRun().Run(utilwait.NeverStop)
+
+	apiserver := kmaster.GenericAPIServer.PrepareRun()
+
+	// presence of the key indicates whether or not to enable the aggregator
+	if len(c.Options.AggregatorConfig.ProxyClientInfo.KeyFile) == 0 {
+		go apiserver.Run(utilwait.NeverStop)
+
+		// Attempt to verify the server came up for 20 seconds (100 tries * 100ms, 100ms timeout per try)
+		cmdutil.WaitForSuccessfulDial(c.TLS, c.Options.ServingInfo.BindNetwork, c.Options.ServingInfo.BindAddress, 100*time.Millisecond, 100*time.Millisecond, 100)
```
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: > Do we not care if we fail here?

pre-existing.  Not really sure what it does

File: pkg/cmd/server/origin/master.go (hunk 6 above)
Comment (enj) [inline pkg/cmd/server/origin/master.go]: > it gets started elsewhere to drive this.

I do not understand what you mean.  All I was saying is that `createAggregatorServer` does not use the `sharedInformers` parameter, so why bother passing it?

File: pkg/cmd/server/origin/master.go (hunk 7 above)
Comment (enj) [inline pkg/cmd/server/origin/master.go]: > pre-existing. Not really sure what it does

Tries a bunch of times and returns and error on failure (which we ignore and just keep going).

**Comment by deads2k** [comment]
Looks like I have some verifies to sort through.

**Comment by smarterclayton** [comment]
This is a blocker for the 3.6 release, I'm putting this at [severity:blocker].  Will review

**Comment by deads2k** [comment]
rebased and fixed up verifies I think

**Comment by deads2k** [comment]
@smarterclayton I seem to be unable to make protobuf verification happy.  Can you try and see?

**Comment by smarterclayton** [comment]
Flake from journalctl output.

**Pull request merged**